### ISSUE TITLE
[Feature Request] Simulations with dynamic bonds based on a pre-computed bond list

### ISSUE THREAD

**Issue by at-ray-o** [openmm/openmm#3731]
Hi, for my project I need to run a polymer simulation where some of the bonds change on the fly. I am only interested in two-body harmonic bonds. I have a pre-computed bond-list as a text file. There is an upper-limit on how many bonds can form at a time, but all particles in the simulation can participate in the bonds. The bonds are not based on distance.

I tried to implement this by running the simulation for some steps and then updating the particle parameter in the bondforce class. However, it seems openmm does not allow the particle parameter to be updated. Is it possible to implement what I want to do?

**Comment by peastman** [comment]
How often do the bonds change?  If you want to change which particles form a bond, you need to call `reinitialize(True)` on the Context.  That's expensive, basically the same as creating a new Context from scratch.  That's fine if they only change infrequently, but not if they could change every few steps.

There might be other ways of accomplishing what you want.  Can you give more details?  How many bonds are present at a time?  How are they determined?  Can any particle be bonded to any other, or are there only certain allowed pairs?

**Comment by at-ray-o** [comment]
Thank you for your response! 
Firstly, the theoretically accurate thing to do would be to change the bonds with every time step, but I am willing to trade some accuracy for speed. It is a little hard for me to say right now how the update frequency will change the accuracy, without trying some simulations. However, if I had to guess a number, I would say about an update after every 100 or so steps would be good.

Secondly, the number of bonds scales linearly with system size. If there are 10,000 particles in the system I am expecting there would be about 2000 bonds. Any particle can be bonded to each other. Which ones are bonded are determined by a kinetic scheme which we calculate separately. The list of pairs of particles bonded are written as a numpy file. I was hoping to read that and update the bonds.

Please let me know if you need more information.

**Comment by peastman** [comment]
I don't really have a better solution than `reinitialize()`.  Do I understand correctly that every 100 steps or so, you will send the current positions to another program (perhaps writing them to a file?), which will compute the new set of bonds, write them to a file, and finally you'll load that file to update the set of bonds?  That process is already going to be pretty slow, so perhaps the reinitialization won't be as disastrously slow as I thought.

Is 10,000 particles typical for what you'll be simulating?  If so, you might try different platforms.  Reinitializing the context is *much* faster for the CPU platform than for CUDA or OpenCL.  It then runs slower, but if you're only working with fairly small systems, it might still end up faster.

**Comment by blake-armstrong** [comment]
I agree with the usefulness of something like this. It would make performing reactive MD more accessible and explorable. Perhaps a method within the `Simulation` object such as `updateTopology(newTopology)` that doesn't require the entire Context to be reinitalised every time a reaction would occur. I personally would be super keen to try and implement MSEVB theory as a means to facilitate some reactive MD into OpenMM! Obviously, this would never be particularly fast through the Python API, but it would be a good place to start.

**Comment by peastman** [comment]
The challenge is to make it efficient.  When you create a Context, one of the things it does is to identify molecules based on the bonds.  The molecules get used for various purposes.  For example, when a barostat scales the periodic box it moves each molecule as a unit rather than scaling each atom individually.  And some platforms internally reorder the atoms to improve computational efficiency, but only in ways that don't disrupt molecules.  If the set of bonds changes, all that information becomes invalid and needs to be redone.

**Comment by at-ray-o** [comment]
Hi, thanks for the insightful comments. @peastman not exactly. All the bonds are pre-computed even before the simulation begins. So I have a file which has all the pairs that I would like to bond at t=0, t=100, t=200 until the end of simulation. I was hoping to read the pairs after every 100 steps and use them to update the bonds. I will try with the `reinitialize()` command and see how it goes. I don't know any gpu coding to suggest how to implement this. I wish I could be of more help in brainstorming that.

**Comment by peastman** [comment]
What is the total number of bonds that exists over the course of the simulation?  If it's small enough, you could include all of them and set their force constants to zero when they're not active.
